add HTTP spec (#508)
libp2p/specs: http/README.md

- Hole punching – Work with peers behind NATs.
- Peer ID Authentication – Authenticate your peer by their libp2p peer id.
- Peer discovery – Learn about a peer given their peer id.

## HTTP Transport vs HTTP Semantics

HTTP is a bit of an overloaded term. This section aims to clarify what we’re talking about when we say “HTTP”.

*HTTP semantics* ([RFC 9110](https://www.rfc-editor.org/rfc/rfc9110.html)) is the stateless application-level protocol that you work with when writing HTTP apis (for example).

*HTTP transport* is the thing that takes your high level request/response defined in terms of HTTP semantics and encodes it and sends it over the wire.

When this document says *HTTP* it is generally referring to *HTTP semantics*.

## Interoperability with existing HTTP systems

A goal of this spec is to allow libp2p to be able to interoperate with existing HTTP servers and clients. Care is taken in this document to not introduce anything that would break interoperability with existing systems.

## HTTP Transport

Nodes MUST use HTTPS (i.e. they MUST NOT use plaintext HTTP). It is RECOMMENDED to use HTTP/2 and HTTP/3.

> Nodes signal support for their HTTP transport using the `/http` component in their multiaddr. e.g. `/dns4/example.com/tls/http` . See the [HTTP multiaddr component spec](https://github.com/libp2p/specs/pull/550) for more details.

Reviewer: :thought_balloon:  so..  this exists:

```console
$ curl -q https://cid.contact/routing/v1/providers/QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR | jq
{
      "Protocol": "unknown",
      "Schema": "unknown",
      "ID": "QmUA9D3H7HeCYsirB3KmPSvZh3dNXMZas6Lwgr4fv1HTTp",
      "Addrs": [
        "/dns4/dag.w3s.link/tcp/443/https"
      ]
    },
```

but https://github.com/libp2p/specs/pull/550 does not mention `/https` notation at all.

@MarcoPolo Should we explicitly state in that spec that `/https` is an alias for `/tls/http`  + do the below here? 
Or was this discussed already elsewhere?

```suggestion
Nodes signal support for their HTTP transport using the `/http` or `/https` component in their multiaddr. e.g. `/dns4/example.com/tls/http` or `/dns4/example.com/https` . See the [HTTP multiaddr component spec](https://github.com/libp2p/specs/blob/master/http/transport-component.md) for more details.
```

Reviewer 2: > Should we explicitly state in that spec that /https is an alias for /tls/http 

`/https` is an alias for `/tls/http`: https://github.com/multiformats/multiaddr/pull/109 and https://github.com/multiformats/multicodec/pull/145

https://github.com/multiformats/multiaddr/blob/master/protocols.csv?plain=1#L31

I'm fine to add a note about `/https` somewhere though

## Namespace

libp2p does not squat the global namespace. libp2p application protocols can be discovered by the [well-known resource](https://www.rfc-editor.org/rfc/rfc8615) `.well-known/libp2p`. This allows server operators to dynamically change the URLs of the application protocols offered, and not hard-code any assumptions how a certain resource is meant to be interpreted.

```json

{

>     "services": {

Reviewer 2: I'm not the biggest fan of the term "services" here. I think I would prefer "protocols", "applications", or maybe something else. Thoughts?

Reviewer: maybe "endpoints"?  (since that is what we define here,  HTTP API endpoints for specific service/protocol)

        "/kad/1.0.0": "/kademlia/",
        "/ipfs-http/1.0.0": "/",
    }
}
```

The resource contains a mapping of application protocols to their respective URL. For example, this configuration file would tell a client

1. That the Kademlia protocol is available at `/kademlia` and
2. The [IPFS Path Gateway API](https://specs.ipfs.tech/http-gateways/path-gateway/) is mounted at `/`.

> It is valid to expose a service at `/`. It is RECOMMENDED that the server resolve more specific URLs before less specific ones. e.g. a path of `/kademlia/foo` should be routed to the Kademlia protocol rather than the IPFS HTTP API.

Reviewer 2: > > It is RECOMMENDED that the server resolve more specific URLs before less specific ones
> 
> Do we need to specify this? Is there an RFC that deals with path resolution that we could link instead?

from @marten-seemann https://github.com/libp2p/specs/pull/556#discussion_r1257365773

---

I believe most HTTP path routing libraries do this already. This is RECOMMENDED because it would be confusing to a client if `/kademlia` (in this example) was routed one way with one server but a different way for another server with the same "services" map.

Reviewer 2: Works for me, but I would change this to a "SHOULD" rather than a "MUST". Thanks!

## Peer ID Authentication

When using the HTTP Transport, peer id authentication is optional. You only pay for it if you need it. This benefits use cases that don’t need peer authentication (e.g. fetching content addressed data) or authenticate some other way (not tied to libp2p peer ids).

Peer ID authentication in the HTTP Transport follows a similar to pattern to how
libp2p adds Peer ID authentication in WebTransport and WebRTC. We run the
standard libp2p Noise handshake, but using `IX` for client and server
authentication or `NX` for just server authentication.

Note: This is just one form of Peer ID authentication. Other forms may be added
in the future (with a different `www-authenticate` value) or be added to the
application protocols themselves.

### Authentication flow

> 1. The client initiates a request that it knows must be authenticated OR the client responds to a `401` with the header `www-authenticate: libp2p-noise` (The server MAY also include `libp2p-token` as an authentication scheme).

Reviewer: nit: other headers follow the canonical form, with each work capitalized:
```suggestion
1. The client initiates a request that it knows must be authenticated OR the client responds to a `401` with the header `WWW-Authenticate: libp2p-noise` (The server MAY also include `libp2p-token` as an authentication scheme).
```

While at it.. I wonder if we should also capitalize `Libp2p-Noise` and `Libp2p-Token` while at it, since preexisting values are `Basic` and `Digest`. Most HTTP software is having normalization code in place for headers anyway, like golang's https://pkg.go.dev/net/http#CanonicalHeaderKey

    3. When the client receives this response, it can authenticate the server’s peer ID.
4. The client verifies the SNI in the Noise extension matches the one used to initiate the connection. The client MUST close the connection if they differ.
    1. The client SHOULD remember this connection is authenticated.
    2. The client SHOULD use the `libp2p-token` if provided for future authorized requests.

This costs one round trip, but can piggy back on an appropriate request.

### Authentication Endpoint

Because the client needs to make a request to authenticate the server, and the client may not want to make the real request before authenticating the server, the server MAY provide an authentication endpoint. This authentication endpoint is like any other application protocol, and it shows up in `.well-known/libp2p`, but it only does the authentication flow. It doesn’t send any other data besides what is defined in the above Authentication flow. The protocol id for the authentication endpoint is `/http-noise-auth/1.0.0`.

## Using HTTP semantics over stream transports

> Application protocols using HTTP semantics can run over any libp2p stream transport. Clients open a new stream using `/http/1.1` as the protocol identifer. Clients encode their HTTP request as an HTTP/1.1 message and send it over the stream. Clients parse the response as an HTTP/1.1 message and then close the stream.

Reviewer 3: This impose strong restriction on the implementations, this makes it way harder to use multiple HTTP clients in one single app.
To speak about what I know about in golang there are two main http implementations used, [`net/http`](https://pkg.go.dev/net/http) and [`fasthttp`](https://pkg.go.dev/github.com/valyala/fasthttp).
Some frameworks like [Fiber](https://gofiber.io/) are only compatible with `fasthttp`.

After talking with @MarcoPolo it seems that the expected minimum API for implementations would be along theses lines:
```go
func (*Host) SetHTTPHandler(protocol.ID, func(remote peer.ID, res ResponseWriter,  req *Request))
```
For equivalence in other languages, that means that libp2p would handle and or delegate deserializing the HTTP requests into a an http request and response writer objects before executing a callback into the consumer's code. This would lock ourself in the `net/http` ecosystem.
Even tho this API is convenient and I do want it in `go-libp2p` I think this spec should target implementation APIs at even lower level where `io.Reader` and `io.Writer` (or equivalent in other programing languages) are used instead.

> Note: Go is a language with an extremely strong std influence, the fact that multiple http libraries are exists and are used in production in Go lead me to question what would happen in a language like Rust or C++ which do not even http libraries in their stds.

Using a single stream makes that way harder to do because you don't know where the request should be routed without parsing it first.
The most reasonable way to solve this for io based routers with a single stream is to write a minimum HTTP parser which parse the path, find out where the request should be routed and then forward the request to the io based consumer with the already parsed buffer and the remaining stream.

However this does not solve a case where I have different applications supporting different versions.
Let's consider the case where I use `h2c` for `/ipfs-gateway-api` (the header compression helps reducing the request size for such a header chatty API or something idk this is an example).
And I also serve `/my-super-application` by a C application which is implemented by using pipes and fork+execve into some handler program, again this is just an example where I would need to use to pass the underying streams.
I can't say to other peers that I'll accept `h2c` but only for `/ipfs-gateway-api`, I have two solutions:
- If I settle for the lowest common denominator. This is bad for my `/ipfs-gateway-api` requests which now run over `http/1.1` and don't get header compression.
- If I advertise `http/h2c` too, what happen if I receive an `h2c` request for `/my-super-application`, I now need to transcode it to `http/1.1` adding lots of complexity.

> Note: I know `h2c` has an upgrade mechanism from `http/1.1` but it does not solve the problem for requests transitioning in the middle of a stream that is being reused (afaik you can't downgrade from `h2c` to `http/1.1` based on the paths you red if you don't like it)

# TL;DR

Using a single stream introduce a whole lot of painful coupling for no reason, we should use a libp2p stream handler per application.

# My proposed solution

This only affects http over stream transports.

## Get rid of  `.well-known/libp2p`

For http over stream transports `.well-known/libp2p` will not be served anywhere, instead the information is passed using the identify protocol.

## Libp2p http streams anouncements

You instead announce things with this format `/http/$PROTOCOL_ID/$HTTP_VERSION` in identify, so my example from previously, I would have theses libp2p stream handlers:
- `/http/ipfs-gateway-api/1.0`
- `/http/ipfs-gateway-api/1.1`
- `/http/ipfs-gateway-api/h2c`
- `/http/my-super-application/1.0`
- `/http/my-super-application/1.1`

For the example earlier in this document you would have thoses libp2p streams being announced:
- `/http/kad/1.0.0/1.0`
- `/http/kad/1.0.0/1.1`
- `/http/ipfs-http/1.0.0/1.0`
- `/http/ipfs-http/1.0.0/1.1`

# Other benefits:

## No weird case in case you wait for identify

This also makes code much simpler for libp2p implementations which make the choice to wait for identify to finish before opening any stream, theses implementations would either
- add a round trip before querying `.well-known/libp2p` because they wait for identify to finish
- add a weird edge case where the HTTP handling code would try to run without identify even tho the implementation might not be equipped to do theses kinds of stream opens.

## It makes it simpler to write libp2p http only implementations

Using `.well-known/libp2p` over libp2p streams is weird, you are emulating libp2p protocol discovery over libp2p, why not just use libp2p's existing infrastructure to do that instead ?
If I am implementing an new libp2p implementation from my app that purely just target http over libp2p instead of also having libp2p over http. I would need to implement the http over libp2p discovery logic even tho I don't support it.
I find that weird.

Reviewer 2: I think the "TL;DR" section is longer than the original section 😉 .

---

> After talking with @MarcoPolo it seems that the expected minimum API for implementations would be along theses lines: ...

Just to be clear, I don't think anything in the document specifies what the API should like for implementations. I also disagree with that example, but let's save that discussion for another thread, maybe in the go-libp2p draft PR that I'll create sometime soon.

On `fasthttp`, there's nothing here that would prevent you from using `fasthttp`. In fact, you can do that today right now https://github.com/libp2p/go-libp2p-http gives you a net.Listener that you can give to `fasthttp.Serve`. Do that and you're done.

On mixing `/h2c` and `/http/1.1` with certain protocols only existing in h2c and others only in 1.1, I don't think this is a top use case we should support. Existing widely deployed HTTP systems don't need this, and we probably don't either. Unless I see a strong compelling use case, I would suggest we ignore this.

On `.well-known/libp2p` and `identify`, you can include the information in `.well-known/libp2p` inside the identify response, so there's no extra cost to this. This isn't spec'd out yet, but should be straightforward.

On `.well-known/libp2p` in general, it's very nice if the way HTTP over a libp2p stream works is identical to the way HTTP over a tls+tcp connection works. It means that a user can pass the same HTTP router (`http.ServeMux` or `http.Handler` in Go) to both an HTTP stack and libp2p stack and it would just work. We already need to define how to route a given path to a given application protocol to work on HTTP, why would we add this again via the protocol id when doing an HTTP over a libp2p stream?

Reviewer 3: The tldr is just:

> Using a single stream introduce a whole lot of painful coupling for no reason, we should use a libp2p stream handler per underlying http service.

---

> On `fasthttp`, there's nothing here that would prevent you from using `fasthttp`. In fact, you can do that today right now https://github.com/libp2p/go-libp2p-http gives you a net.Listener that you can give to `fasthttp.Serve`. Do that and you're done.

https://github.com/libp2p/go-libp2p-http only concerns the client not the server AFAIK.

---

> On mixing /h2c and /http/1.1 with certain protocols only existing in h2c and others only in 1.1, I don't think this is a top use case we should support. Existing widely deployed HTTP systems don't need this, and we probably don't either. Unless I see a strong compelling use case, I would suggest we ignore this.

I don't understand your point, using a libp2p stream per protocol cost nothing and I IMO simplify the code because you do not have to do two services resolution (first identify then http wellknown) I can see the http wellknown being equal argument but this is only valid if you assume absolutely all libp2p http implementations would want to do both plain http and or http over libp2p streams which I don't belive to be true.

---

> On `.well-known/libp2p` and `identify`, you can include the information in `.well-known/libp2p` inside the identify response, so there's no extra cost to this. This isn't spec'd out yet, but should be straightforward.

Why not specify it now ? My spec is easy (`/http/$PROTOCOL_ID/$HTTP_VERSION`) and will not require weird backward compatible code where the information is both in identify and `.wellknown/libp2p` ?

---

> (`http.ServeMux` or `http.Handler` in Go) to both an HTTP stack and libp2p stack and it would just work.

No, what if I am using fiber which uses fasthttp ? I wont be compatible with the `http.ServeMux` nor `http.Handler`.
Go is a best case scenario, what about rust where there is no clear standard HTTP lib ?
It is not urealistical for them to have an actix handler and an hyper handler in the same app.
Using different stream handlers for different libp2p http apps cost nothing while not doing so has real costs in edge cases.

Reviewer 3: Having everything on one single `/http/1.1` stream handler is more complex than it looks like, because with pipelinable protocols like `http/1.1` the first request can be done on application handler 1 which is actix based while the second request could be handled on a different handler backed by hyper.

Reviewer 2: There is no pipelining here. This document says each HTTP request and response happen in a single stream. You use a libp2p stream per request.

Reviewer 3: Actually my proposal only works with http over libp2p not libp2p over http so it's value is really limited.

> HTTP/1.1 is chosen as the minimum bar for interoperability, but other encodings of HTTP semantics are possible as well and may be specified in a future update.

Reviewer 3: I think the `minimum bar for interoperability` should be `HTTP/1.0`, given libp2p streams are multiplexed you shouldn't need to reuse them.
This gives a good hint to servers that they can skip the overhead of chunked encoding and just close the underlying libp2p stream.

Reviewer: Yes, +1 that we should require 1.1  – it is the minimal bar these days with updated  RFC and brings support for more optional things than just chunking: 
- Caching: Cache-Control and ETag
- Host header for virtual hosting
- Pipelining  and persistent connections (multiple requests and responses to be sent over a single connection)

Reviewer 3: Sorry I'm not suggesting you should use HTTP1.0 but you should remove the extra no-pipelining restrictions.

Reviewer 2: Let’s continue discussion on https://github.com/libp2p/specs/pull/508#discussion_r1525604274

Interest Group: [@lidel]

[@marten-seemann]: https://github.com/marten-seemann
[@MarcoPolo]: https://github.com/MarcoPolo
[@lidel]: https://github.com/lidel

## Introduction

This document defines how libp2p nodes can offer and use an HTTP transport alongside their other transports to support application protocols with HTTP semantics. This allows a wider variety of nodes to participate in the libp2p network, for example:

- Browsers communicating with other libp2p nodes without needing a WebSocket, WebTransport, or WebRTC connection.
- HTTP only edge workers can run application protocols and respond to peers on the network.
- `curl` from the command line can make requests to other libp2p nodes.

As well as allowing application protocols to make use of HTTP intermediaries such as HTTP caching and layer 7 proxying and load balancing. This is all in addition to the existing features that libp2p provides such as:

- Connectivity – Work on top of WebRTC, WebTransport, QUIC, TCP, or an HTTP transport.
- Hole punching – Work with peers behind NATs.
- Peer ID Authentication – Authenticate your peer by their libp2p peer id.
- Peer discovery – Learn about a peer given their peer id.

> ## HTTP Transport vs HTTP Semantics

Reviewer 4: I don't know if there is precedent for this in libp2p specs, but I thought the visual diagram you had created for earlier HTTP discussions was useful.

Reviewer 2: Hesitant to include that diagram since it's a png and hard to update. I've tried replicating with Mermaid so that it's text based. wdyt? I don't think it's a nice and maybe a bit more confusing than the text.

Reviewer 4: I've attached the image I was referring to so others are aware.

It seems like that diagram was particularly useful when it came to talking about how "vanilla/simple request response" maps.  That's not in scope for this spec.  I'm still not opposed to having the diagram you have, although I do think it would be useful to color code the lines like you did in the diagram below to cover the "HTTP over libp2p" and "Just HTTP" cases.

![Libp2p request response](https://github.com/libp2p/specs/assets/85411/dd5a8718-7d7e-4032-b045-faba0d0c6c80)

HTTP is a bit of an overloaded term. This section aims to clarify what we’re talking about when we say “HTTP”.

*HTTP semantics* ([RFC 9110](https://www.rfc-editor.org/rfc/rfc9110.html)) is the stateless application-level protocol that you work with when writing HTTP apis (for example).

*HTTP transport* is the thing that takes your high level request/response defined in terms of HTTP semantics and encodes it and sends it over the wire.

When this document says *HTTP* it is generally referring to *HTTP semantics*.

## Interoperability with existing HTTP systems

> A goal of this spec is to allow libp2p to be able to interoperate with existing HTTP servers and clients. Care is taken in this document to not introduce anything that would break interoperability with existing systems.

Reviewer 2: Refers to both actually

## HTTP Transport

Nodes MUST use HTTPS (i.e. they MUST NOT use plaintext HTTP). It is RECOMMENDED to use HTTP/2 and HTTP/3.

Nodes signal support for their HTTP transport using the `/http` component in their multiaddr. e.g. `/dns4/example.com/tls/http` . See the [HTTP multiaddr component spec](https://github.com/libp2p/specs/pull/550) for more details.

## Namespace

libp2p does not squat the global namespace. libp2p application protocols can be discovered by the [well-known resource](https://www.rfc-editor.org/rfc/rfc8615) `.well-known/libp2p`. This allows server operators to dynamically change the URLs of the application protocols offered, and not hard-code any assumptions how a certain resource is meant to be interpreted.

```json

{
    "services": {
        "/kad/1.0.0": "/kademlia/",

>         "/ipfs-http/1.0.0": "/",

Reviewer 5: Some more thoughts on this:

1. The path alone is not enough to form a request. We need more (out-of-band) information like what method to use.
2. Do you expect a protocol-spec like kademlia to be extended on how it can be accessed from an HTTP transport? (And which methods need authentication for example?)

>         "/ipfs-http/1.0.0": "/",

Reviewer 2: I'm a bit hesitant to add the ability to put more metadata without a use case behind it, but making it a map with path seems easy enough, and would allow adding new fields in a backwards compatible way, which is nice.

---

> 1. The path alone is not enough to form a request. We need more (out-of-band) information like what method to use.

This is up to the application protocol. The application protocol defines how it works and what HTTP methods it uses for what. This metadata only describes where the application protocol is mounted at. It doesn't describe the application protocol.

> 2. Do you expect a protocol-spec like kademlia to be extended on how it can be accessed from an HTTP transport? (And which methods need authentication for example?)

Yes. That's up to the application protocol. Kademlia could make use of the simple request-response abstraction in #561 to define this, but ultimately the application protocol decides tis.

>         "/ipfs-http/1.0.0": "/",

Reviewer 2: I've updated this to be a map from protocol name to metadata that includes a path.

>         "/ipfs-http/1.0.0": "/",

Reviewer 5: > > 3\. Do you expect a protocol-spec like kademlia to be extended on how it can be accessed from an HTTP transport? (And which methods need authentication for example?)
> 
> Yes. That's up to the application protocol. Kademlia could make use of the simple request-response abstraction in #561 to define this, but ultimately the application protocol decides tis.

I see, thanks for explaining. Can we make this clearer? Something like:

This mapping only defines a URL namespace for certain applications protocols. It is entirely up to the application protocol (like kademlia) to define how it can be interacted with over HTTP.

### Authentication flow

1. The client initiates a request that it knows must be authenticated OR the client responds to a `401` with the header `www-authenticate: libp2p-noise` (The server MAY also include `libp2p-token` as an authentication scheme).
2. The client sets the `Authorization` [header](https://www.rfc-editor.org/rfc/rfc9110.html#section-11.6.2) to `libp2p-noise <multibase-encoded-noise-protobuf>` . This initiates the `IX` or `NX` handshake.
    1. The protobuf is multibase encoded, but clients MUST only use encodings that are HTTP header safe (refer to to the [token68 definition](https://www.rfc-editor.org/rfc/rfc9110.html#section-11.2)). To set the minimum bar for interoperability, clients and servers MUST support base32 encoding (”b” in the multibase table).
    2. When the server receives this request and `IX` was used, it can authenticate the client.
3. The server responds with `Authentication-Info` field set to `libp2p-noise <multibase-encoding-noise-protobuf-response>`.
    1. The server MUST include the SNI used for the connection in the [Noise extensions](https://github.com/libp2p/specs/blob/master/noise/README.md#noise-extensions).

>     2. The server MAY include a token that the client can use to avoid doing another Noise handshake in the future. The client would use this token by setting the `Authorization` header to `libp2p-token <token>`.

Reviewer 2: Cookie headers have a bit of baggage with them. For example they are a [Forbidden header name](https://developer.mozilla.org/en-US/docs/Glossary/Forbidden_header_name), and cannot be used with a service worker proxy.

The token would be included in the Noise extensions along with the SNI used. That wasn't clear, thanks for pointing that out!

    3. When the client receives this response, it can authenticate the server’s peer ID.
4. The client verifies the SNI in the Noise extension matches the one used to initiate the connection. The client MUST close the connection if they differ.
    1. The client SHOULD remember this connection is authenticated.
    2. The client SHOULD use the `libp2p-token` if provided for future authorized requests.

This costs one round trip, but can piggy back on an appropriate request.

### Authentication Endpoint

Because the client needs to make a request to authenticate the server, and the client may not want to make the real request before authenticating the server, the server MAY provide an authentication endpoint. This authentication endpoint is like any other application protocol, and it shows up in `.well-known/libp2p`, but it only does the authentication flow. It doesn’t send any other data besides what is defined in the above Authentication flow. The protocol id for the authentication endpoint is `/http-noise-auth/1.0.0`.

> ## Using HTTP semantics over stream transports

Reviewer 5: So do we also define how to map protocols to an HTTP transport? I am a bit confused. If not, why do we need the whole mapping of protocol ID to endpoint then?

Reviewer 2: > So do we also define how to map protocols to an HTTP transport?

This document is only concerned about application protocols written with HTTP semantics. The mapping of protocol ID to path is to tell a peer where a certain HTTP-semantic application protocol exists on the server. For example, I could have Trustless gateway mounted on `/` and an s3-compatible api on `/s3` and Kademlia on `/kad`. Kademlia would be "HTTP semantic aware" even if it did this by using #561 .

Reviewer 5: If we address https://github.com/libp2p/specs/pull/508#discussion_r1267030234, this is resolved for me.

# HTTP

| Lifecycle Stage | Maturity      | Status | Latest Revision |
| --------------- | ------------- | ------ | --------------- |
| 1A              | Working Draft | Active | r0, 2023-01-23  |

Authors: [@marten-seemann], [@MarcoPolo]

Interest Group: [@lidel], [@thomaseizinger]

[@marten-seemann]: https://github.com/marten-seemann
[@MarcoPolo]: https://github.com/MarcoPolo
[@lidel]: https://github.com/lidel
[@thomaseizinger]: https://github.com/thomaseizinger

## Introduction

This document defines how libp2p nodes can offer and use an HTTP transport alongside their other transports to support application protocols with HTTP semantics. This allows a wider variety of nodes to participate in the libp2p network, for example:

- Browsers communicating with other libp2p nodes without needing a WebSocket, WebTransport, or WebRTC connection.
- HTTP only edge workers can run application protocols and respond to peers on the network.
- `curl` from the command line can make requests to other libp2p nodes.

> The HTTP transport will also allow application protocols to make use of HTTP intermediaries such as HTTP caching, and layer 7 proxying and load balancing. This is all in addition to the existing features that libp2p provides such as:

Reviewer 6: Related to https://github.com/libp2p/specs/pull/508/files#r1541421460 but not all HTTP methods are equally cachable/proxyable - should the spec have an opinion on how requests are sent?

Reviewer 6: On balance, it'll probably be application specific.

Reviewer 2: Applications will interact with HTTP semantics, so they will have directly control how requests are sent.

Applications may choose to use helper libraries that can map some other request/response protocol onto HTTP semantics. In that case, that library defines the mapping. This spec itself only suggests one way of doing that with `POST`, but other ways are also valid.

    HTTP1.1 --- HTTPTransport
    HTTP2 --- HTTPTransport
    HTTP3 --- HTTPTransport
```

- *HTTP semantics* ([RFC 9110](https://www.rfc-editor.org/rfc/rfc9110.html)) is
  the stateless application-level protocol that you work with when writing HTTP
  apis (for example).

- *HTTP encoding* is the thing that takes your high level request/response
  defined in terms of HTTP semantics and encodes it into a form that can be sent
  over the wire.

- *HTTP transport* is the thing that takes your encoded reqeust/response and
  sends it over the wire. For HTTP/1.1 and HTTP/2, this is a TCP+TLS connection.
  For HTTP/3, this is a QUIC connection.

When this document says *HTTP* it is generally referring to *HTTP semantics*.

## Interoperability with existing HTTP systems

A goal of this spec is to allow libp2p to be able to interoperate with existing HTTP servers and clients. Care is taken in this document to not introduce anything that would break interoperability with existing systems.

## HTTP Transport

Nodes MUST use HTTPS (i.e., they MUST NOT use plaintext HTTP). It is RECOMMENDED to use HTTP/2 and HTTP/3.

Nodes signal support for their HTTP transport using the `/http` component in

> their multiaddr. E.g., `/dns4/example.com/tls/http`. See the [HTTP multiaddr

Reviewer 6: Peer ID authentication is optional, but if you have to add a PeerID in the sample multiaddr where does it fit?

There has been a lot of discussion about HTTP multiaddrs being path multiaddrs in the past, which would prevent `/p2p/Qmfoo` being added onto the end, though I don't think they were ever resolved.

Reviewer 2: I have an open issue for how to document HTTP resources: https://github.com/libp2p/specs/issues/554 and a corresponding PR: https://github.com/multiformats/multiaddr/pull/164

With that change it would be very much like it is today:
`/dns4/example.com/tls/http/p2p/Qmfoo` or `/dns4/example.com/tls/http/httppath/somepath/p2p/Qmfoo`

This is consistent with interpreting multiaddrs [right to left](https://github.com/multiformats/multiaddr/?tab=readme-ov-file#interpreting-multiaddrs), as we first see the peer we are communicating with, the (optional) HTTP resource we are referencing, and the direction to connect to the peer.

Reviewer 6: Makes sense 👍 

https://github.com/multiformats/multiaddr/pull/164 suggests adding the `httppath` component after the Peer ID, which seems to contradict the above.  Maybe it needs updating?

> `/httppath` should be appended to the end of an existing multiaddr, including after the peer id component (p2p). As an example, here's a multiaddr referencing the `.well-known/libp2p` HTTP resource along with a way to reach that peer:
>
> ```
> /ip4/1.2.3.4/tcp/443/tls/http/p2p/12D.../httppath/.well-known%2Flibp2p
> ```

Source: https://github.com/multiformats/multiaddr/pull/164/files#diff-940635fb4ac11927020564dd7f96ac0604da98bb11aee1273030de75a31dcf06R27

Reviewer 2: > https://github.com/multiformats/multiaddr/pull/164 suggests adding the httppath component after the Peer ID, which seems to contradict the above. Maybe it needs updating?

Good catch. Maybe my past self was right and my current self is wrong. I know my past self spent a lot more time thinking about that document than my current self did in writing that comment.

Reviewer 2: My past self was right. and my comment [above](https://github.com/libp2p/specs/pull/508#discussion_r1525335920) is wrong.

Multiaddrs are interpreted right to left, but the peer-id is more about *where* this resource is located. We first learn about the HTTP resource we're interested in, and then learn about how to get it.

component spec](https://github.com/libp2p/specs/pull/550) for more details.

## Namespace

libp2p does not squat the global namespace. libp2p application protocols can be
discovered by the [well-known resource](https://www.rfc-editor.org/rfc/rfc8615)

> `.well-known/libp2p`. This allows server operators to dynamically change the

Reviewer 6: We use well-known resources [elsewhere](https://docs.libp2p.io/concepts/transports/webtransport/#webtransport-in-libp2p), e.g. `.well-known/libp2p-webtransport`.  Perhaps we should namespace this one too?

```suggestion
`.well-known/libp2p-http`. This allows server operators to dynamically change the
```

Reviewer 2: Aesthetically it's nicer if we aren't referencing HTTP multiple times. `.well-known/libp2p` is a HTTP resource. It's kind of like saying "ATM machine". 

Technically this doesn't matter, but my vote is for `.well-known/libp2p`

Reviewer 6: I'm not sure I agree. Given we have other well-known resources, `.well-known/libp2p` is ambiguous, `.well-known/libp2p-http` is not.

Though yes, from a technical perspective it just has to be a predictable string with a low chance of collision.

Reviewer 2: @lidel can you be our tie breaker? `.well-known/libp2p` vs `.well-known/libp2p-http`

Reviewer: Squatting `.well-known/libp2p` for a file may cause us problems in the future if we need to add something else.

I think we should go with either `.well-known/libp2p-http` ..or make a directory and put protocol mapping configuration in  `.well-known/libp2p/http`.

The latter (`libp2p/http`) feels a bit better to me, as it creates libp2p-specific directory/namespace, which is easier to map via reverse proxies, but other than that it is mostly aesthetics.

Reviewer 2: Just a note that this is a map, so we could always add more things in a backwards compatible way.

I like namespacing with `.well-know/libp2p`. This lets us get a single IANA URI suffix if we want while still giving us flexibility in what goes under it.

I'll make the change to `.well-known/libp2p/protocols`, thanks. Mark this one as another success for IPFS' Chief Renaming Officer.

Reviewer 6: Just to point out that we use [.well-known/libp2p-webtransport](https://docs.libp2p.io/concepts/transports/webtransport/#webtransport-in-libp2p) for WebTransport things already, `.well-known/libp2p-http` would be consistent with that, `.well-known/libp2p/protocols` looks like something different.

Reviewer 2: This is different though. This is metadata about a peer's supported protocols. The well-know webtransport URI is about where to send the HTTP CONNECT request to.

Reviewer 6: I'm not sure it is, I think how you interact with the well-known resource is a detail of the protocol that's unrelated to the path it uses.

Reviewer 2: The `.well-known/libp2p/` suffix has the benefit of being a single thing we would need to register. Future versions of libp2p-webtransport may be placed under that suffix. Remember WebTransport isn't even out of draft status yet. We'll probably need to make a new multiaddr for webtransport-v1 just like we did with QUIC. Any other well-known resource we would want could also fit under that suffix.

Reviewer 7: Changing this from "/.well-known/libp2p" to "/.well-known/libp2p/protocols" will break the ability to communicate with existing servers deployed without upgrading them first. Since these servers are running in places not under our control, they will not be immediately upgradable.

Because there are already deployments that use the old value, I suggest keeping it as `/.well-known/libp2p`

Reviewer 2: I understand the pain in this rollout. Especially when the full rollout is outside your control. Maybe go-libp2p can provide a transition period to help out?

Since this spec itself is in draft we need flexibility in how things work. As much as I prefer the `.well-known/libp2p`, I don't think the argument of "we should use it because we've deployed this draft" is a strong one. Deployments of drafts are extremely useful to get some experience, but we shouldn't ossify to those decisions because we already made them.

Once this is merged, things will be stable and we won't break existing users. That's the guarantee that comes with merging this. And partly the reason why I haven't merged it yet. I want to build the js-libp2p side before merging and make sure that the interop works and is reasonable. I'm pretty close on that. Hoping to get it done by the end of the month :)

        "/ipfs/gateway": {"path": "/"},
    }
}
```

The resource contains a mapping of application protocols to a URL namespace. For
example, this configuration file would tell a client

1. That the Kademlia application protocol is available with prefix `/kademlia`
and,
2. The [IPFS Trustless Gateway API](https://specs.ipfs.tech/http-gateways/trustless-gateway/) is mounted at `/`.

It is valid to expose a service at `/`. It is RECOMMENDED that implementations facilitate the coexistence of different service endpoints by ensuring that more specific URLs are resolved before less specific ones. For example, when registering handlers, more specific paths like `/kademlia/foo` should take precedence over less specific handler, such as `/`.

## Peer ID Authentication

When using the HTTP Transport, peer id authentication is optional. You only pay
for it if you need it. This benefits use cases that don’t need peer
authentication (e.g., fetching content addressed data) or authenticate some
other way (not tied to libp2p peer ids).

Specific authentications schemes for authenticating Peer IDs will be defined in
a future spec.

## Using HTTP semantics over stream transports

> Application protocols using HTTP semantics can run over any libp2p stream transport. Clients open a new stream using `/http/1.1` as the protocol identifer. Clients encode their HTTP request as an HTTP/1.1 message and send it over the stream. Clients parse the response as an HTTP/1.1 message and then close the stream.

Reviewer 2: ```suggestion
Application protocols using HTTP semantics can run over any libp2p stream transport. Clients open a new stream using `/http/1.1` as the protocol identifer. Clients encode their HTTP request as an HTTP/1.1 message and send it over the stream. Clients parse the response as an HTTP/1.1 message and then close the stream. Clients MUST NOT pipeline requests over a single stream.
```

Reviewer 3: Why `MUST` ? There are valid use-cases for pipelining (when the responses only make sense in order).
It can be surprisingly hard to make existing HTTP implementation not reuse streams.

Reviewer 2: This is only when using HTTP semantics on top of libp2p streams so existing implementations don't apply here. Right?

> There are valid use-cases for pipelining (when the responses only make sense in order).

Maybe, but it also complicates things. As a key datapoint HTTP/3 does not support request pipelining: "A client MUST send only a single request on a given stream." from https://datatracker.ietf.org/doc/html/rfc9114#name-expressing-http-semantics-i . This is consistent with that.

Reviewer 3: > This is only when using HTTP semantics on top of libp2p streams so existing implementations don't apply here. Right?

I would expect nearly all libp2p-http implementations to use an existing http implementation like you did in `go-libp2p/p2p/http`.
I see your implementation does in fact not reuse stream however that because it's not compliant with `http1.1` you don't implement transfer encoding nor harden against miss-match of `Content-Length` and body (\*that what I found in 5 minutes of looking).
Your protocol does make sense as a simpler thing but then you can't call it `HTTP1.1` it's a subset of `HTTP1.1` where only closing the stream is used to indicate EOF (akin to HTTP1.0 but with `Host` headers).

The way I would have written `go-libp2p/p2p/http` is instead of implementing a custom HTTP transport, I would have used `http.Transport` with a custom `DialContext` handler and most of the features turned off (TLS, HTTP2, ...) aside of correctness, it also allow features like timeouts `req.Context`, `Accept-Encoding` for compression and others to work.

Reviewer 2: > The way I would have written go-libp2p/p2p/http is instead of implementing a custom HTTP transport, I would have used http.Transport with a custom DialContext handler 

I think this would force you to serialize all HTTP requests, which is not a great default.

What's the use case that would benefit most from allowing pipelining? Can it not be addressed without pipelining?

Reviewer 3: > I think this would force you to serialize all HTTP requests.

What do you mean ?

---

> What's the use case that would benefit most from allowing pipelining? Can it not be addressed without pipelining?

It helps resumptions of trustless gateway downloads.
You are in the middle of a file, you have a DFS stack of CIDs you need.
First I ask you for the inner most part of the dag I don't have, then the next one, then the next one.

If I do that without pipelining, I end up doing one round-trip for every unprocessed CID in the dag, which can easily be 50k+ if downloading files produced by optimized implementations.
Even with kubo defaults you can have to send thousands or so round trips in large ish files. (it gets worst as the number of links per block increase)

Instead of asking each CID one by one, I can use request pipelining to request for the next CIDs I want before I received the first ones.
The usual HTTP wisdom would be that this is bad because if a request stalls it blocks all the requests behind, but in this case I'm streaming the file I very precisely don't want the chunks out of order.

The alternatives are using multiple streams, which has worst resources control because the server will try to send me multiple blocks in parallel which I don't actually care about.
For this very particular usecase I could also recompute a single trustless gateway request which spans the undownloaded region of the files, however this is always gonna send me duplicate root blocks I already have, this is not future proof.

We don't gain anything server side, HTTP1.1 always makes it legal for the server to use EOF as indicating the end of a response (by not using `Content-Length` nor `Transfer-Encoding`).

Reviewer 2: I'm not convinced by this use case.

First, it's important to point out that the **only** reason for pipelining in HTTP/1.1 is because you could not make a concurrent request without a whole new TCP connection. HTTP/2 and HTTP/3 dropped pipelining support because they introduced better ways to make concurrent requests.

This general problem of "I have a list of things I want to fetch in order" (which I believe is the same thing as your trustless gateway example) is not best solved through request pipelining. It is convenient to use a single channel and dump all your requests at the beginning, but there are downsides with this solution:

1. Request pipelining is not part of HTTP semantics. Therefore the application protocol should make no assumption on whether or not its request will be pipelined. The same application should run the same whether it's running on top of HTTP/1.1, h2, or h3. 
1. The request may fail at any point. This means there is more complicated logic besides "This request failed or suceeded." It also means you can't offload state to the request pipeline.
1. Sometimes the list of things you need changes (You seek somewhere else in the file), cancelling a single requests is a lot more straightforward than cancelling a whole stack of requests.
1. It's possible (and likely) that a set of concurrent request will be addressed faster than a set of serial requests, even if the client only cares about the responses in order. Imagine a 3/4 blocks are cached, but the last 1/4 needs to be fetched from a spinning disk.

The only thing that would convince me that we should support pipelining is if it would make implementations much simpler. My current experience with Go and JS is that that is not the case. And my brief survey into Rust leads me to think it's not necessary there either.

Not supporting pipelining for HTTP/1.1 encoding over libp2p streams does simplify implementations. One example: The library can safely close streams once it has sent a response. We avoid issues where the server closed a stream, but the client sent another request on that stream before it saw the close.

Reviewer 2: I'm also going to add that the server SHOULD send back a `Connection: close` header. That should help enforce the idea that these connections over streams are not persistent.

Reviewer 3: 1. That looks like a me (client) problem, I don't see why the spec should disallow this.
2. I don't see why this is a problem, reset the stream and rematerialize the new request state from the unixfs DFS stack, I precisely don't want to track individual requests because a request might "fail" half way through but still send me useful blocks I was able to verify.
3. It's not, reset the whole stream and the whole stack is canceled. Then do rematerialization.
4. I agree it's faster, but it's less efficient, more buffer space is used, the server likely allocate more stacks, ... the RTT from the client to the server is often multiple orders worst than RTT from server to it's disks.
  This is way more tricky to implement efficiently, I can't double buffer it as if the blocks are small I'm not gonna fill up the pipe, so I need to track some concurrent requests window in my application layer. While with the pipelining solution I have a read and a write goroutine, and `.Write` is gonna block when the pipe is full, backpressure provided by libp2p.

> Not supporting pipelining for HTTP/1.1 encoding over libp2p streams does simplify implementations. One example: The library can safely close streams once it has sent a response. We avoid issues where the server closed a stream, but the client sent another request on that stream before it saw the close.

That not simplifying anything, the stream reusing is already an optional part of HTTP1.1, if you don't want to implement it send `Connection: close` and don't do it (https://github.com/libp2p/go-libp2p/pull/2733 aimed at proving this).

Reviewer 2: If I change this from a MUST to a SHOULD would that alleviate your concerns here? I'm not convinced pipelining is a good thing here, but I can understand the desire that if a specific server and client want to do it, the spec shouldn't prohibit it. Server implementations can (SHOULD in the spec's term) send a `Connection: close` to not reuse streams.

Reviewer 3: Yes thx

> Application protocols using HTTP semantics can run over any libp2p stream transport. Clients open a new stream using `/http/1.1` as the protocol identifer. Clients encode their HTTP request as an HTTP/1.1 message and send it over the stream. Clients parse the response as an HTTP/1.1 message and then close the stream. Clients SHOULD NOT pipeline requests over a single stream. Servers SHOULD set the [`Connection: close` header](https://datatracker.ietf.org/doc/html/rfc2616#section-14.10) to signal to clients that this is not a persistent connection.

Reviewer 8: NIT: 
https://datatracker.ietf.org/doc/html/rfc9112#section-9.6

Reviewer 8: Should we add a line that explains that a libp2p stream is equivalent of a TCP connection for http/1.1?

Reviewer 2: There might be implications with associating a stream to a TCP connection. I don't want to think about all of those, so I'll leave it as is.

HTTP/1.1 is chosen as the minimum bar for interoperability, but other encodings of HTTP semantics are possible as well and may be specified in a future update.

## Using other request-response semantics (not HTTP)

This document has focused on using HTTP semantics, but HTTP may not be the common divisor amongst all transports (current and future). It may be desirable to use some other request-response semantics for your application-level protocol, perhaps something like rust-libp2p’s [request-response](https://docs.rs/libp2p/0.52.1/libp2p/request_response/index.html) abstraction. Nothing specified in this document prohibits mapping other semantics onto HTTP semantics to keep the benefits of using an HTTP transport.

> To support the simple request-response semantics, for example, the request MUST be encoded within a `POST` request to the proper URL (as defined in the Namespace section). The response is read from the body of the HTTP response. The client MUST authenticate the server and itself **before** making the request.

Reviewer 6: `POST` makes caching & load balancing (a stated goal of this transport) hard because it is not idempotent.

Maybe prefer `PUT` here?

Reviewer 2: It is intended to not use caching. HTTP caching is tied to HTTP semantics, if you want caching you should build on HTTP semantics. This is not trying to make existing protocols magically get caching for free.

Reviewer 2: Added a sentence to this effect.

> To support the simple request-response semantics, for example, the request MUST be encoded within a `POST` request to the proper URL (as defined in the Namespace section). The response is read from the body of the HTTP response. The client MUST authenticate the server and itself **before** making the request.

Reviewer 6: > The client MUST authenticate the server and itself **before** making the request.

Are there any thoughts on how this will work?

There's a limitation here that when using `fetch` in JS, the whole request must be sent before any response bytes are read, that means you can't do an authentication preamble.

You can read the response headers before the rest of the response, but you've already sent your data by this point, which might have been very large.

  defined in terms of HTTP semantics and encodes it into a form that can be sent
  over the wire.

- *HTTP transport* is the thing that takes your encoded reqeust/response and
  sends it over the wire. For HTTP/1.1 and HTTP/2, this is a TCP+TLS connection.
  For HTTP/3, this is a QUIC connection.

When this document says *HTTP* it is generally referring to *HTTP semantics*.

## Interoperability with existing HTTP systems

A goal of this spec is to allow libp2p to be able to interoperate with existing HTTP servers and clients. Care is taken in this document to not introduce anything that would break interoperability with existing systems.

## HTTP Transport

Nodes MUST use HTTPS (i.e., they MUST NOT use plaintext HTTP). It is RECOMMENDED to use HTTP/2 and HTTP/3.

Nodes signal support for their HTTP transport using the `/http` component in
their multiaddr. E.g., `/dns4/example.com/tls/http`. See the [HTTP multiaddr
component spec](https://github.com/libp2p/specs/pull/550) for more details.

## Namespace

libp2p does not squat the global namespace. libp2p application protocols can be
discovered by the [well-known resource](https://www.rfc-editor.org/rfc/rfc8615)

> `.well-known/libp2p/protocols`. This allows server operators to dynamically change the

Reviewer: I'd say it is good practice to keep it.
Makes it easy to quickly validate JSON, allows us to add more things to the file, if the need arises, without breaking legacy clients

URLs of the application protocols offered, and not hard-code any assumptions how
a certain resource is meant to be interpreted.

```json

{
    "protocols": {
        "/kad/1.0.0": {"path": "/kademlia/"},
        "/ipfs/gateway": {"path": "/"},
    }
}
```

The resource contains a mapping of application protocols to a URL namespace. For
example, this configuration file would tell a client

1. That the Kademlia application protocol is available with prefix `/kademlia`
and,

> 2. The [IPFS Trustless Gateway API](https://specs.ipfs.tech/http-gateways/trustless-gateway/) is mounted at `/`.

Reviewer: Hm.. the methods will be specific to each protocol at each mount point, so not part of this spec?

> 2. The [IPFS Trustless Gateway API](https://specs.ipfs.tech/http-gateways/trustless-gateway/) is mounted at `/`.

Reviewer 8: That makes sense. As I see it there are two sections in this document: 
1. Running libp2p protocols over standard http like h2 or h3
2. Running http protocols over libp2p streams.

So should we mention it in the specs that a libp2p protocol supporting http transport should specify the http method and headers to be used for the protocol. For the path they can expose it via the wellknown endpoint `/.well-known/libp2p/protocols`

> 2. The [IPFS Trustless Gateway API](https://specs.ipfs.tech/http-gateways/trustless-gateway/) is mounted at `/`.

Reviewer 2: I'm not sure I understand. An application protocol would be built using HTTP semantics, and that protocol would then be able to run on libp2p streams or "standard" http transports like h2, h3.

What do you mean by:
> Running libp2p protocols over standard http like h2 or h3

This spec does not define how you would take an existing libp2p protocol and map it to HTTP semantics. That is best done by the specific protocol itself. But maybe I'm misunderstanding your point?

It is valid to expose a service at `/`. It is RECOMMENDED that implementations facilitate the coexistence of different service endpoints by ensuring that more specific URLs are resolved before less specific ones. For example, when registering handlers, more specific paths like `/kademlia/foo` should take precedence over less specific handler, such as `/`.

## Peer ID Authentication

When using the HTTP Transport, peer id authentication is optional. You only pay
for it if you need it. This benefits use cases that don’t need peer
authentication (e.g., fetching content addressed data) or authenticate some
other way (not tied to libp2p peer ids).

Specific authentications schemes for authenticating Peer IDs will be defined in
a future spec.

## Using HTTP semantics over stream transports

> Application protocols using HTTP semantics can run over any libp2p stream transport. Clients open a new stream using `/http/1.1` as the protocol identifer. Clients encode their HTTP request as an HTTP/1.1 message and send it over the stream. Clients parse the response as an HTTP/1.1 message and then close the stream. Clients SHOULD NOT pipeline requests over a single stream. Servers SHOULD set the [`Connection: close` header](https://datatracker.ietf.org/doc/html/rfc9112#section-9.6) to signal to clients that this is not a persistent connection.

Reviewer 8: Is there a way for a multiaddr to signal support for this?
 
In most cases it doesn't make sense to speak http/1.1 over quic-v1 as specified in this section because you can simply do h3 over quic-v1 which is represented by the multiaddr:
`/ip4/1.2.3.4/udp/42/quic-v1/http`

Is there a way to specify that the node also speaks http/1.1(the variant specified in this section) over quic libp2p streams?

# HTTP

| Lifecycle Stage | Maturity      | Status | Latest Revision |
| --------------- | ------------- | ------ | --------------- |
| 1A              | Working Draft | Active | r0, 2023-01-23  |

Authors: [@marten-seemann], [@MarcoPolo]

Interest Group: [@lidel], [@thomaseizinger]

[@marten-seemann]: https://github.com/marten-seemann
[@MarcoPolo]: https://github.com/MarcoPolo
[@lidel]: https://github.com/lidel
[@thomaseizinger]: https://github.com/thomaseizinger

## Introduction

> This document defines how libp2p nodes can offer and use an HTTP transport alongside their other transports to support application protocols with HTTP semantics. This allows a wider variety of nodes to participate in the libp2p network, for example:

Reviewer 9: This will be my last comment. Expending effort on using HTTP is a signal that moves the center of gravity of the project — verbiage of scope or bindingness does not change this. The desire to make HTTP requests is best met by the programmer using an HTTP client library. You have not answered how a browser shall successfully make an HTTPS request to a peer, given that non-cloud peers will not be able to present valid X.509 certificates.

- `curl` from the command line can make requests to other libp2p nodes.

The HTTP transport will also allow application protocols to make use of HTTP intermediaries such as HTTP caching, and layer 7 proxying and load balancing. This is all in addition to the existing features that libp2p provides such as:

- Connectivity – Work on top of WebRTC, WebTransport, QUIC, TCP, or an HTTP transport.
- Hole punching – Work with peers behind NATs.
- Peer ID Authentication – Authenticate your peer by their libp2p peer id.
- Peer discovery – Learn about a peer given their peer id.

## HTTP Semantics vs Encodings vs Transport

HTTP is a bit of an overloaded term. This section aims to clarify what we’re talking about when we say “HTTP”.

```mermaid
graph TB
    subgraph "HTTP Semantics"
        HTTP
    end
    subgraph "Encoding"
        HTTP1.1[HTTP/1.1]
        HTTP2[HTTP/2]
        HTTP3[HTTP/3]
    end
    subgraph "Transports"
        Libp2p[libp2p streams]
        HTTPTransport[HTTP transport]
    end
    HTTP --- HTTP1.1
    HTTP --- HTTP1.1
    HTTP1.1 --- Libp2p
    HTTP --- HTTP2
    HTTP --- HTTP3
    HTTP1.1 --- HTTPTransport
    HTTP2 --- HTTPTransport

>     HTTP3 --- HTTPTransport

Reviewer 3: h2c to show how it can be multiplexed and negociated in many ways (header compression, binary based protocol, ...) ?
Feel free to ignore this comment if you think it's making the graph too complex.

## Peer ID Authentication

When using the HTTP Transport, peer id authentication is optional. You only pay for it if you need it. This benefits use cases that don’t need peer authentication (e.g., fetching content addressed data) or authenticate some other way (not tied to libp2p peer ids).

Peer ID authentication in the HTTP Transport follows a similar to pattern to how
libp2p adds Peer ID authentication in WebTransport and WebRTC. We run the
standard libp2p Noise handshake, but using `IX` for client and server
authentication or `NX` for just server authentication.

Note: This is just one form of Peer ID authentication. Other forms may be added
in the future (with a different `WWW-Authenticate` value) or be added to the
application protocols themselves.

### Authentication flow

1. The client initiates a request that it knows must be authenticated OR the client responds to a `401` with the header `WWW-Authenticate: Libp2p-Noise` (The server MAY also include `Libp2p-Token` as an authentication scheme).

> 2. The client sets the `Authorization` [header](https://www.rfc-editor.org/rfc/rfc9110.html#section-11.6.2) to `Libp2p-Noise <multibase-encoded-noise-protobuf>` . This initiates the `IX` or `NX` handshake.

Author: We might run into the HTTP header size limit here (e.g. for HTTP/3: https://datatracker.ietf.org/doc/html/rfc9114#section-4.2.2), especially when using PQ keys.

Would it make sense to use a separate HTTP endpoint here, and do the handshake using the body of a request / response to that endpoint?

Reviewer 2: How big are pq keys? It's pretty nice and ergonomic to do auth in the headers because then standard stuff like this works:

```go
func(w http.ResponseWriter, r *http.Request) {
    // This handler requires libp2p auth
    clientID, err := AuthenticateClient(serverKey, w.Header(), r.Header)
    
    if err != nil || clientID == "" {
	w.Header().Set("WWW-Authenticate", "Libp2p-Noise-IX")
	w.WriteHeader(http.StatusUnauthorized)
	return
    }
    
    w.WriteHeader(http.StatusOK)
    w.Write([]byte(fmt.Sprintf("Hello, %s", clientID)))
}
```

Or even middlewares:

```go
libp2pAuthMiddleware(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
	w.WriteHeader(http.StatusOK)
	w.Write([]byte(fmt.Sprintf("Hello, %s", r.Context().Value(libp2pAuthKey))))
}))
```

A [survey](https://www.geekersdigest.com/max-http-request-header-size-server-comparison/) shows that many support at least 8KB in headers.

We could only support the libp2p-token in the auth header, and do the the noise handshake on a dedicated endpoint. That would mean we wouldn't be able to piggyback on a request to do auth. It would also mean that doing something with client auth is always a 2 step process. It also adds a bit of complexity I think since there are two flows for auth, instead of everything happening in the headers.

If we can stay below the 8KB limit, [which may be possible](https://blog.cloudflare.com/sizing-up-post-quantum-signatures/) I think sticking with the headers for everything would be nicer.

    1. The protobuf is multibase encoded, but clients MUST only use encodings that are HTTP header safe (refer to to the [token68 definition](https://www.rfc-editor.org/rfc/rfc9110.html#section-11.2)). To set the minimum bar for interoperability, clients and servers MUST support base32 encoding (”b” in the multibase table).
    2. When the server receives this request and `IX` was used, it can authenticate the client.
3. The server responds with `Authentication-Info` field set to `Libp2p-Noise <multibase-encoding-noise-protobuf-response>`.
    1. The server MUST include the SNI used for the connection in the [Noise extensions](https://github.com/libp2p/specs/blob/master/noise/README.md#noise-extensions).
    2. The server MAY include a token in the Noise extensions that the client
    can use to avoid doing another Noise handshake in the future. The client
    would use this token by setting the `Authorization` header to `Libp2p-Token

>     <token>`.

Reviewer 5: Don't we have to define the code point of this token? I am not too familiar with the design intentions around our noise extension registry but from an implementer PoV, I wouldn't know at the moment, how to extract this token.

Reviewer 2: Yep, I just pushed the relevant changes to the noise spec. It adds two new optional pb fields

# libp2p + HTTP: the spec

| Lifecycle Stage | Maturity                 | Status | Latest Revision |
|-----------------|--------------------------|--------|-----------------|
| 1A              | Working Draft            | Active | r0, 2023-01-23  |

Authors: [@marten-seemann]

Interest Group: [@MarcoPolo]

[@marten-seemann]: https://github.com/marten-seemann
[@MarcoPolo]: https://github.com/MarcoPolo

## Introduction

This document defines how libp2p nodes can offer a HTTP endpoint next to (or instead of) their full libp2p node. Services can be offered both via traditional libp2p protocols and via HTTP, allowing a wide variety of nodes to access these services. Crucially, this for the first time, allows browsers to access libp2p services without spinning up a Web{Socket, Transport, RTC} connection first. It also allows interacting with libp2p services from environments where plain HTTP is the only option, e.g. curl from the command line, and certain cloud edge workers and lambdas.

At the same time, nodes that are already connected via a libp2p connection, will be able to (re)use this connection to issue the same kind of requests, without dialing a dedicated HTTP connection.

> Any protocol that follows request-response semantics can easily be mapped onto HTTP (mapping protocols that don’t follow a request-response flow can be more challenging). Protocols are encouraged to follow best practices for building REST APIs. Once a mapping has been defined, a single implementation can be used to serve both traditional libp2p as well as libp2p-HTTP clients.

Reviewer 2: What is a mapping?

## Addressing

Nodes may advertise HTTP multiaddresses to signal support for libp2p over HTTP. An address might look like this: `/ip4/1.2.3.4/tcp/443/tls/sni/example.com/http/p2p/<peer id>` (for HTTP/1.1 and HTTP/2), or  `/ip4/1.2.3.4/udp/443/quic/sni/example.com/http/p2p/<peer id>` (for HTTP/3).

Nodes MUST use HTTPS (i.e. they MUST NOT use unencrypted HTTP). It is RECOMMENDED to use HTTP/2 and HTTP/3, but the protocols also work over HTTP/1.1.

> Note that the peer ID in this address is 1. optional and 2. advisory and not (necessarily) verified during the HTTP handshake (depending on the HTTP client). If and when desired, clients can cryptographically verify the peer ID once the HTTP connection has been established, see [Authentication] for details on peer authentication.

Author: > This is essentially the user asking for server peer id authentication.

Who's the user here? Client or server?

I think I agree in general. Maybe we should forbid `/p2p` when using HTTP, unless we find a way to _always_ verify the peer ID? Encoding a signature into the subdomain could be a way to do that.

Nodes can also link to a specific resource directly, similar to how a URL includes a path. This will require us to resolve [https://github.com/multiformats/multiaddr/issues/63](https://github.com/multiformats/multiaddr/issues/63) first. For example, the URL of a specific CID might be: `/ip4/1.2.3.4/tcp/443/tls/sni/example.com/http/<my data transfer protocol>/{/path/to/<cid>}`.

## Namespace

libp2p does not squat the global namespace. By convention, all libp2p services are located at a well-known URL: `http://example.com/.well-known/libp2p/<service name>/<path (optional)>`.

Putting the service name into the URL allows for future extensibility. It is easy to define new protocols, and the replace existing protocols by newer versions.

> Applications MAY expose services under different URIs. For example, an application might decide to generate nicer-looking (and probably more SEO-friendly) URLs, and map paths under `[https://example.com/dht/](https://example.com/dht/)` to `https://example.com/.well-known/libp2p/kad-dht-v1/`. 

Author: 1. We need to have HTTP endpoints for libp2p-internal protocols. This includes Identify (or a split-up version of that protocol, as @thomaseizinger suggested), Kademlia, Gossipsub (?), etc. At the very least, we need to decide where these protocols are available.
2. Who gets to decide where protocols are mapped? A typical stack will have multiple layers of protocols. If they all start polluting the global namespace, it won't be possible to run a normal webserver any more, since the URL endpoints will be taken. In my view, _only_ the top layer should make decisions about the global namespace.

> Applications MAY expose services under different URIs. For example, an application might decide to generate nicer-looking (and probably more SEO-friendly) URLs, and map paths under `[https://example.com/dht/](https://example.com/dht/)` to `https://example.com/.well-known/libp2p/kad-dht-v1/`. 

Reviewer 5: We could leverage links here:

1. Protocols define paths. Let's say, kademlia has the `/kademlia/values` path and accepts `PUT` requests there.
2. Applications can mount the paths of a libp2p library anywhere they want.
3. We define an index document that MUST sit in a well-known location, i.e. `example.com/.well-known/libp2p`.
4. Said index document lays out, where to find the libp2p paths (and maybe other metadata):

```json
{
	"version": "0.25.0",
	"mounted_at": "/libp2p"
}
```
5. Implementations could then discover the mount point and automatically prefix the `/libp2p` segment. The kademlia module would then make requests to `/libp2p/kademlia/values`.

This way, the global namespace is not squatted and protocols at the same time have freedom over their URL design.

> Applications MAY expose services under different URIs. For example, an application might decide to generate nicer-looking (and probably more SEO-friendly) URLs, and map paths under `[https://example.com/dht/](https://example.com/dht/)` to `https://example.com/.well-known/libp2p/kad-dht-v1/`. 

Reviewer 5: An alternative to `./well-known/libp2p` would be to tell applications to include a `Link` header in their index response. Implementations would then:

1. Make a `HEAD` request to `/`
2. Search for a `Link` header with `rel=libp2p`, i.e. `Link: /libp2p; rel="libp2p"`

Even if said application hosts many many different websites and endpoints, it shouldn't be too much to ask to include a `Link` header in a GET/HEAD request to `/`.

> Applications MAY expose services under different URIs. For example, an application might decide to generate nicer-looking (and probably more SEO-friendly) URLs, and map paths under `[https://example.com/dht/](https://example.com/dht/)` to `https://example.com/.well-known/libp2p/kad-dht-v1/`. 

Reviewer 5: If I remember correctly, then per web linking spec (RFC 8288), you cannot just invent link relations and they should be registered with IANA instead so we can't just use `libp2p`. But, the spec also defines what is called an extended link relation. Those are links (inception anyone?) that MAY resolve to a document explaining the link relation. Something like:

`Link: </libp2p-sub-path>; rel="https://docs.libp2p.io/link-relations/mount-point"`

Going to https://docs.libp2p.io/link-relations/mount-point could then show a piece of documentation that explains what implementations should do here.

> Applications MAY expose services under different URIs. For example, an application might decide to generate nicer-looking (and probably more SEO-friendly) URLs, and map paths under `[https://example.com/dht/](https://example.com/dht/)` to `https://example.com/.well-known/libp2p/kad-dht-v1/`. 

Reviewer 5: All of this is at the cost of 1 round-trip when you first talk to a given libp2p node, assuming we cache the response.

It is valid to expose a service at `/`. It is RECOMMENDED that implementations facilitate the coexistence of different service endpoints by ensuring that more specific URLs are resolved before less specific ones. For example, when registering handlers, more specific paths like `/kademlia/foo` should take precedence over less specific handler, such as `/`.

## Peer ID Authentication

When using the HTTP Transport, peer id authentication is optional. You only pay for it if you need it. This benefits use cases that don’t need peer authentication (e.g., fetching content addressed data) or authenticate some other way (not tied to libp2p peer ids).

Peer ID authentication in the HTTP Transport follows a similar to pattern to how
libp2p adds Peer ID authentication in WebTransport and WebRTC. We run the
standard libp2p Noise handshake, but using `IX` for client and server
authentication or `NX` for just server authentication.

Note: This is just one form of Peer ID authentication. Other forms may be added
in the future (with a different `WWW-Authenticate` value) or be added to the
application protocols themselves.

### Authentication flow

> 1. The client initiates a request that it knows must be authenticated OR the client responds to a `401` with the header `WWW-Authenticate: Libp2p-Noise-IX` (The server MAY also include `Libp2p-Token` as an authentication scheme).

Author: For all headers we use here, we should define them as structured fields (RFC 8941). This will allow us to use standardized header parsers (e.g. https://github.com/dunglas/httpsfv for Go) do the parsing for us, instead of writing a bespoke parser for every field we define.

There might be some objection to not pulling in multibase here, but I think that using an IETF-specified way to do header encoding outweighs the benefits of multibase.

Reviewer 2: I'm not opposed. It looks fairly straightforward and it seems like there are libraries out for JS, Rust, and Go at least.

But, to be clear, we aren't doing bespoke parsing. We're using protobufs here, and we encode it into a multibase. This is just like what we do everywhere else we need to send structured data over the wire.

With the current definition, SFV might not get us much since we still need to reach for a protobuf parser to read the [NoiseHandshake](https://github.com/libp2p/specs/tree/master/noise#the-libp2p-handshake-payload) payload and we need Multihash to parse the peer id. So it could only replace the multibase encoding. Changing "b<base32>" (or "m<base64>")  to ":<base64>:".

We could consider switching to a different auth scheme, but I think that's a bigger change, and I wanted to avoid introducing a new auth scheme.

> To speed of server authentication, a node MAY include the libp2p TLS extension in its certificate. Note that this is currently not possible when using Let’s Encrypt, since the libp2p TLS extension is not whitelisted by LE. Not every HTTP client will have access to the TLS certificate (for example, browsers usually don’t expose an API for that), but if an HTTP client does, it SHOULD use that information.

Author: Currently there isn't any, as far as I'm aware. Once we have a strong use case and a spec, we could start reaching out to CAs to see if it's possible to be included in their whitelist.

> Since HTTP requests are independent from each other (they are not bound to a single connection, and when using HTTP/1.1, will actually use different connections), the server needs to authenticate itself on every single request.

Author: Maybe? On the other hand, if you're willing to do a TLS handshake for every HTTP/1.1 request, maybe you don't really care about a signature. Of course this will look different for a HTTP/2 or HTTP/3 connection...
When caching, you'd have to be very careful about HTTP redirects.

There's also the risk that this signature scheme negates the benefits of using a CDN: Unless the libp2p key is at the edge, every request has to go back to the origin to perform this signature.

> Since HTTP requests are independent from each other (they are not bound to a single connection, and when using HTTP/1.1, will actually use different connections), the server needs to authenticate itself on every single request.

Author: > these things together mean the responses for the same thing sent to different users will have different header, and will not be reusable across users. Effectively making it impossible to benefit from HTTP caching across users.

Is that so? HTTP headers shouldn't have any effect on cacheability, unless you include them in the [Vary](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Vary) header.

> For example, DHT lookup for a peer records would hit the backend server every time, even tho you could safely cache results for a while at the edge CDN.

The first question I'd ask is if it makes sense to request authentication in that case. I imagine that _almost all_ requests we handle won't need peer authentication, since we're dealing with either signed or content-addressed data anyway.

> To request the server to authenticate, the client sets the `libp2p-server-auth` HTTP header to a randomly generated ASCII string of at least 10 (and a maximum of 100) characters. The server signs the following string using its host key:

Author: That sounds reasonable.

> To request the server to authenticate, the client sets the `libp2p-server-auth` HTTP header to a randomly generated ASCII string of at least 10 (and a maximum of 100) characters. The server signs the following string using its host key:

Reviewer 2: Resolving since this has changed significantly.

> 1. `libp2p-server-pubkey`: its public key (from the libp2p key pair)

Reviewer 2: The update changes this significantly

> When requesting server authentication, the client MUST check that these two header fields are present, and MUST check the signature. It MUST NOT process the response if either one of these checks fails

Reviewer 5: Maybe it is too obvious but the client must also check that the domain matches the one it sent the request to!

Reviewer 5: In other words, the client should rebuild the above string from its own knowledge, i.e. use the domain it sent the request to and check that the signature matches against this value.

Given the variable length nature of domains and the client-controlled ASCII string, perhaps we should add a hashing step too?

Reviewer 2: I'm wondering if we should use JWT here as well.

Reviewer 2: Resolving since we've changed this significantly. It should answer the original question.

> ### Client Authentication

Reviewer 2: Can we use https://webauthn.guide/#about-webauthn here instead of making our own protocol? I'd rest easier if we leveraged existing work here. I need to think about tradeoffs some more (will it be annoying outside of the browser context?), but generally want to avoid custom auth schemes if we can.

Reviewer 5: That would be fantastic.

Reviewer 2: We still can, but that will be out of scope for this spec. This spec will outline *a* way of doing auth, not *the* way. Other ways will be supported.

> In order to keep this exchange stateless, the server SHOULD 1. include the current timestamp or an expiry data and 2. a signature in that data. This allows it to check in step 2 that it actually generated that data.

Reviewer 2: I'm a little confused. Are you saying that the server needs to include the expiry data and their signature as part of the `random` field? Are you suggesting they put the useful state in the field and encrypt it? And that's the `random` field?

Author: Yes, this basically offloads that state to the client, in a way that the client can't manipulate.

Services SHOULD specify the caching properties (if any), and set the appropriate cache headers (according to RFC 9111).

CDNs can also be used to increase censorship resistance, since the CDN effectively hides the IP address of the origin server. With the upcoming introduction of ECHO (Encrypted ClientHello) in TLS, all that an on-path observer will be able to see is that a client is establishing a connection to a certain CDN, but not to which domain name.

The level of delegation between the origin node and the CDN can be adjusted. In the simplest configuration, the origin node is the only node that holds the libp2p private key, thus requests to the `server-auth` protocol would be forwarded from the CDN to the origin server. In a more advanced configuration, it would be possible to move the private key to a worker on the edge of the CDN, and perform the signing operation there (thereby reducing the request latency for `server-auth` requests).

## FAQ

### Why not gRPC?

> This would be the perfect fit, allowing both request-response schemes as well as variations with multiple requests and multiple responses. However, it’s not possible to use gRPC from the browser.

Reviewer 5: I am very confused by this wording in the document. gRPC has no built-in semantics and just shifts bytes back and forth. We already have that today with our bi-directional streams.

Isn't the whole point of this idea to move away from those "dumb" streams and instead define protocols using HTTP and its rich semantics instead? We are not going to get any of the benefits I listed above if we end up with gRPC.

AFAIK, gRPC is only defined on top of HTTP because that makes it easy to deploy (firewalls allowing port 80 etc).

> This would be the perfect fit, allowing both request-response schemes as well as variations with multiple requests and multiple responses. However, it’s not possible to use gRPC from the browser.

Reviewer 2: This section has been cut

Nodes can also link to a specific resource directly, similar to how a URL includes a path. This will require us to resolve [https://github.com/multiformats/multiaddr/issues/63](https://github.com/multiformats/multiaddr/issues/63) first. For example, the URL of a specific CID might be: `/ip4/1.2.3.4/tcp/443/tls/sni/example.com/http/<my data transfer protocol>/{/path/to/<cid>}`.

## Namespace

libp2p does not squat the global namespace. By convention, all libp2p services are located at a well-known URL: `http://example.com/.well-known/libp2p/<service name>/<path (optional)>`.

Putting the service name into the URL allows for future extensibility. It is easy to define new protocols, and the replace existing protocols by newer versions.

Applications MAY expose services under different URIs. For example, an application might decide to generate nicer-looking (and probably more SEO-friendly) URLs, and map paths under `[https://example.com/dht/](https://example.com/dht/)` to `https://example.com/.well-known/libp2p/kad-dht-v1/`. 

### Service Names

> Traditionally, libp2p protocols have used path-like protocol identifiers, e.g. `/libp2p/autonat/1.0.0`. Due to the use of `/`s, this doesn’t work well with the naming convention defined above.

Author: Technically, the `%2F` encoding makes it a valid URL component, doesn’t it? ;)

Given that protocols need to opt-in into using HTTP anyway, and that will require changes to their spec (see https://github.com/libp2p/specs/pull/508#issuecomment-1399418780 for my reasoning), I thought it wouldn’t be too much to ask for to define a (shorter, URL-friendlier, human-readable) name. Not a super strong opinion though.

> Traditionally, libp2p protocols have used path-like protocol identifiers, e.g. `/libp2p/autonat/1.0.0`. Due to the use of `/`s, this doesn’t work well with the naming convention defined above.

Reviewer 2: Resolving since I've updated the spec and I think this has been covered

The protocol defined here takes 2 RTTs to authenticate the client. It is designed to be stateless on the server side. In the first round-trip, the client obtains a (pseudo-) random value from the server, which it then signs with its host key and sends back to the server, which then issues an authentication token (acting somewhat like a cookie) which can be included on future requests.

The service name is `client-auth`. For the first step, the client sends a GET request to this HTTP endpoint, and the server responds with at least 8 and up to 1024 bytes of pseudorandom data:

```json
{
	"random": <multibase-encoded random bytes>
}
```

In order to keep this exchange stateless, the server SHOULD 1. include the current timestamp or an expiry data and 2. a signature in that data. This allows it to check in step 2 that it actually generated that data.

The client signs the data received in step 1, and sends a POST request with the following JSON object to the server:

```json
{
  "data": <the random bytes received from the server, multibase encoded>,
  "peer-id": <peer ID, in string reprensentation>,
  "signature": <multibase-encoded signature>
}
```

> The server verifies the signature and issues an authentication token. In order to allow stateless operation, at the very minimum, the authentication token SHOULD contain the peer ID. It SHOULD also contain an expiry date and it MAY be bound to the client’s IP address. The token is sent in the response body.

Reviewer 5: Perhaps JWT could serve us well here?

https://www.rfc-editor.org/rfc/rfc7519

Author: I thought of that. In the end, I decided to not include it here since it doesn’t matter how the server encodes the value, as long as it’s able to decode it when the client sends the token. JWT seems like a reasonable choice.

On the other hand, there might be value in recommending (or maybe even requiring?) the use of JWT, as it would make the interaction with intermediaries easier: A server could share an encryption key (which can, but doesn’t have to be the host key) with a CDN, allowing the decoding of the JWT at the edge.

Reviewer 2: I like the JWT approach since it seems like we could define one less thing and reuse the JWT libraries already out there.

Reviewer 2: Resolving since I've updated the spec and I think this has been covered

The client uses the auth token on requests that require client authentication, by setting the `libp2p-auth-token` HTTP header.

## Mapping to libp2p Streams

libp2p services whose service is specified as request-response protocols can use a single protocol implementation to make the service available over HTTP as well as on top of libp2p streams.

> The libp2p protocol identifier is `/http1.1`. After negotiating this protocol using multistream-select, nodes treat the stream as a HTTP/1.1 stream for a single HTTP request (i.e. nodes MUST NOT use request pipelining).

Reviewer 2: This is how we do HTTP over the existing stream abstraction. This allows applying everything above on top of an existing stream (if we have one).

This is useful in WebTransport or WebRTC scenarios where you may be talking to a node that doesn't have a valid TLS certificate but still supports the protocol you want to use.

## Outlook: Interaction with Intermediaries

One of the advantages of running HTTP is that there’s widely deployed caching infrastructure (CDNs). Content-addressed data is infinitely cacheable. Assuming a properly design data transfer protocol, retrieval for CIDs could be cached by the CDN and made available via a POP (geographically) close to the user, dramatically reducing retrieval latencies.

> Services SHOULD specify the caching properties (if any), and set the appropriate cache headers (according to RFC 9111).

Author: Yes, in my understanding, the request method is part of the specification for a REST endpoint, isn’t it?

> Services SHOULD specify the caching properties (if any), and set the appropriate cache headers (according to RFC 9111).

Reviewer 2: Protocols will use the existing HTTP abstractions to define their logic.

This is different than rust-libp2p's [request-response protocol](https://docs.rs/libp2p/0.50.0/libp2p/request_response/index.html). Instead of a custom serialization of request/response, we adopt all of the HTTP semantics.

Specifically, using libp2p+HTTP will allow:

1. Defining services / protocols once, and run them both via HTTP and via libp2p
1. Leverage libp2p's connectivity story (incl. hole punching) to run these services on both public nodes and on nodes behind NATs / firewall
1. Use existing peer and content discovery mechanisms to advertise HTTP-enabled multiaddresses, which can then be accessed either via plain HTTP(S) or via HTTP on top of libp2p
    1. Support existing HTTP protocols like the S3 protocol. This would allow peers to fetch content seamlessly from an S3-compatible provider (S3, backblaze's B2, Cloudflare's R2)
    1. Support edge compute directly. Many edge compute environments build on top of HTTP since it’s a stateless request/response protocol. This includes services such as Cloudflare workers, AWS Lambda, Netflify Edge functions, and many more.
1. Use peer authentication (both client and server auth) for a subset of HTTP endpoints

## Addressing

> Nodes may advertise HTTP multiaddresses to signal support for libp2p over HTTP. An address might look like this: `/ip4/1.2.3.4/tcp/443/tls/sni/example.com/http/p2p/<peer id>` (for HTTP/1.1 and HTTP/2), or  `/ip4/1.2.3.4/udp/443/quic/sni/example.com/http/p2p/<peer id>` (for HTTP/3).

Reviewer 10: (1)
Do we have any concerns about the fact that the `http` segment is not followed with another value, as seems to be the case for all other segments (dns, ipv4, tcp, p2p)? I am not familiar with the multiaddr spec, but if we always have segments in the form `/{key}/{value}`, then parsers don't need to understand all key values to be able to parse the multiaddr into a list of parts. If some keys are followed by a value and some are not, then introducing a new value-less key requires changes in all parsers to be aware of this new key.

(2) 
What would a DNS-based multiaddr look like? When deploying libp2p nodes to the cloud, the IP addresses are often dynamic and change over time; only the hostname is always the same.

I have a js-libp2p node deployed to Fly.io at https://saturn-link-poc.fly.dev ([GitHub repo](https://github.com/bajtos/saturn-interop-libp2p)), accepting libp2p connections at the following multiaddr:
```
/dns/saturn-link-poc.fly.dev/tcp/3030/p2p/12D3KooWRH71QRJe5vrMp6zZXoH4K7z5MDSWwTXXPriG9dK8HQXk
```

Let's say I want to enable support for HTTP. What would be the new address? I am assuming something like this:

```
/dns/saturn-link-poc.fly.dev/tcp/80/sni/saturn-link-poc.fly.dev/http/p2p/12D3KooWRH71QRJe5vrMp6zZXoH4K7z5MDSWwTXXPriG9dK8HQXk
```

It's a long address with some redundancy. Can we avoid duplicating the hostname in the `dns` and `sni` parts?

Ideally, I would want to write:

```
/dns/saturn-link-poc.fly.dev/tcp/80/http/p2p/12D3KooWRH71QRJe5vrMp6zZXoH4K7z5MDSWwTXXPriG9dK8HQXk
```

Author: > (1)
> Do we have any concerns about the fact that the `http` segment is not followed with another value, as seems to be the case for all other segments (dns, ipv4, tcp, p2p)? I am not familiar with the multiaddr spec, but if we always have segments in the form `/{key}/{value}`, then parsers don't need to understand all key values to be able to parse the multiaddr into a list of parts. If some keys are followed by a value and some are not, then introducing a new value-less key requires changes in all parsers to be aware of this new key.

No need to worry here, all parsers should understand these already. For example, both `/quic` and `/p2p-circuit` don't have an argument.

> Let's say I want to enable support for HTTP. What would be the new address? I am assuming something like this:
> 
> ```
> /dns/saturn-link-poc.fly.dev/tcp/80/sni/saturn-link-poc.fly.dev/http/p2p/12D3KooWRH71QRJe5vrMp6zZXoH4K7z5MDSWwTXXPriG9dK8HQXk
> ```

Yes, the `/sni` component is optional. If you're using a DNS name, you don't need to specify it.

> Ideally, I would want to write:
> 
> ```
> /dns/saturn-link-poc.fly.dev/tcp/80/http/p2p/12D3KooWRH71QRJe5vrMp6zZXoH4K7z5MDSWwTXXPriG9dK8HQXk
> ```

That's exactly how it would look like. Except, you probably want to use HTTPS, so it would be
```
/dns/saturn-link-poc.fly.dev/tcp/443/tls/http/p2p/12D3KooWRH71QRJe5vrMp6zZXoH4K7z5MDSWwTXXPriG9dK8HQXk
```

## Addressing

Nodes may advertise HTTP multiaddresses to signal support for libp2p over HTTP. An address might look like this: `/ip4/1.2.3.4/tcp/443/tls/sni/example.com/http/p2p/<peer id>` (for HTTP/1.1 and HTTP/2), or  `/ip4/1.2.3.4/udp/443/quic/sni/example.com/http/p2p/<peer id>` (for HTTP/3).

Nodes MUST use HTTPS (i.e. they MUST NOT use unencrypted HTTP). It is RECOMMENDED to use HTTP/2 and HTTP/3, but the protocols also work over HTTP/1.1.

Note that the peer ID in this address is 1. optional and 2. advisory and not (necessarily) verified during the HTTP handshake (depending on the HTTP client). If and when desired, clients can cryptographically verify the peer ID once the HTTP connection has been established, see [Authentication] for details on peer authentication.

Nodes can also link to a specific resource directly, similar to how a URL includes a path. This will require us to resolve [https://github.com/multiformats/multiaddr/issues/63](https://github.com/multiformats/multiaddr/issues/63) first. For example, the URL of a specific CID might be: `/ip4/1.2.3.4/tcp/443/tls/sni/example.com/http/<my data transfer protocol>/{/path/to/<cid>}`.

## Namespace

> libp2p does not squat the global namespace. By convention libp2p services can be discovered by accessing the configuration file at a well-known URL (RFC 8615): `.well-known/libp2p.json` (e.g. https://example.com/.well-known/libp2p.json). This allows server operators to dynamically change the URLs of the services offered, and to not hard-code any assumptions how a certain resource is meant to be interpreted.

Reviewer 2: Also there still discussion around having `.well-known/libp2p.json` end with `.json` vs having it defined in the Accept https://github.com/libp2p/specs/pull/529

> libp2p does not squat the global namespace. By convention libp2p services can be discovered by accessing the configuration file at a well-known URL (RFC 8615): `.well-known/libp2p.json` (e.g. https://example.com/.well-known/libp2p.json). This allows server operators to dynamically change the URLs of the services offered, and to not hard-code any assumptions how a certain resource is meant to be interpreted.

Reviewer 2: Not necessary, so no need to tie ourselves to this. We need to check `.well-known` in the general case anyways so there isn't much benefit to canonical paths.
